satpy/composites/__init__.py

```diff
@@ -1714,24 +1726,51 @@ def _combine_metadata_with_mode_and_sensor(self,
 
     @staticmethod
     def _get_merged_image_data(foreground: xr.DataArray,
-                               background: xr.DataArray
+                               background: xr.DataArray,
+                               initial_bg_alpha: bool,
                                ) -> list[xr.DataArray]:
-        if "A" in foreground.attrs["mode"]:
-            # Use alpha channel as weight and blend the two composites
-            alpha = foreground.sel(bands="A")
-            data = []
-            # NOTE: there's no alpha band in the output image, it will
-            # be added by the data writer
-            for band in foreground.mode[:-1]:
-                fg_band = foreground.sel(bands=band)
-                bg_band = background.sel(bands=band)
-                chan = (fg_band * alpha + bg_band * (1 - alpha))
-                chan = xr.where(chan.isnull(), bg_band, chan)
-                data.append(chan)
-        else:
-            data_arr = xr.where(foreground.isnull(), background, foreground)
-            # Split to separate bands so the mode is correct
-            data = [data_arr.sel(bands=b) for b in data_arr["bands"]]
+        def _get_alpha(dataset: xr.DataArray):
+            # If the dataset contains an alpha channel, just use it
+            # If not, we still need one. So build it and fill it with 1
+            if "A" in dataset.attrs["mode"]:
+                alpha = dataset.sel(bands="A")
+            else:
+                first_band = dataset.isel(bands=0)
+                alpha = xr.full_like(first_band, 1)
+                alpha["bands"] = "A"
+
+            # There could be Nans in the alpha, especially through 'add_bands'
+            # Replace them with 0 to prevent cases like 1 + nan = nan, so they won't affect new_alpha
+            alpha = xr.where(alpha.isnull(), 0, alpha)
+
+            return alpha
+
+        alpha_fore = _get_alpha(foreground)
+        alpha_back = _get_alpha(background)
+        new_alpha = alpha_fore + alpha_back * (1 - alpha_fore)
+
+        data = []
+
+        # Unless background has an initial alpha band, there will be no alpha band in the output image
+        # Let the writer decide
+        output_mode = background.attrs["mode"] if initial_bg_alpha else background.attrs["mode"].replace("A", "")
+
+        # If we let the writer decide alpha band, we must fill the transparent areas in the image with np.nan first
+        # The best way is through the new alpha
+        new_alpha_nan = xr.where(alpha_fore + alpha_back == 0, np.nan, new_alpha) if "A" not in output_mode \
+            else new_alpha
```

Comment:
Could some of this logic and maybe most of `_get_alpha` be removed if we were more clever with the `bands` argument that is passed to `add_bands` earlier in the compositor? For example, if instead of passing `backgrounds["bands"]` to `add_bands` we figured out what the final `output_mode` was going to be and passed that to `add_bands` instead, then we wouldn't have to do some of this `xr.where` call to create a new alpha band...I hope.

What I'm really saying is that it is really hard for me to follow what this method is doing. That may just be me or it may be the complicated logic being implemented here or it might be that there is room for improvement in the code.

Reply:
I think `add_bands` requires both the band data and the mode attribute. I don't know how to do it by just passing mode.

Reply:
I do think it would be possible to move some of this up, but given how much I see this logic mirrors the wikipedia article then maybe that isn't for the best.